scripts/globals/battlefield.lua

```diff
@@ -68,10 +70,788 @@ xi.battlefield.dropChance =
+    obj.delayToExit = data.delayToExit or 5 -- Amount of time to wait before exiting the battlefield. Defaults to 5 seconds.
+    obj.groups = {} -- Monster battlefield groups added with battlefield:addGroups()
+    obj.paths = {} -- Pathing for monsters and npcs within the battlefield
+    obj.loot = {} -- Loot spawned in the Armoury Crate(s)
+    obj.requiredItems = data.requiredItems or {} -- Items required to be traded to enter the battlefield
+    obj.requiredKeyItems = data.requiredKeyItems or {} -- Key items required to be able to enter the battlefield - these are removed upon entry
+    obj.entryNpc = data.entryNpc -- The name of the NPC used for entering
+    obj.exitNpc = data.exitNpc -- The name of the NPC used for exiting
+    obj.title = data.title -- Title given to players upon victory
+    obj.grantXP = data.grantXP -- Amount of XP to grant upon victory
+    obj.lossEventParams = data.lossEventParams or {} -- Parameters given to the loss event (32002). Defaults to none.
+    obj.sections = { { [obj.zoneId] = {} } }
+    return obj
+end
+
+function Battlefield.getVarPrefix(battlefieldID)
+    return string.format("Battlefield[%d]", battlefieldID)
+end
+
+function Battlefield:register()
+    -- Only hookup the entry and exit listeners if there aren't any other battlefields already registered for that entrance
+    local setupEvents = true
+    local setupEntryNpc = true
+    local setupExitNpc = true
+    if utils.hasKey(self.zoneId, xi.battlefield.contentsByZone) then
+        local contents = xi.battlefield.contentsByZone[self.zoneId]
+        for _, content in ipairs(contents) do
+            -- Always setup listeners if we're reloading a battlefield
+            if self.battlefieldId == content.battlefieldId and content.hasListeners then
+                setupEvents = true
+                setupEntryNpc = true
+                setupExitNpc = true
+                break
+            end
+
+            -- Do not setup events if there are any other battlefields
+            setupEvents = false
+
+            -- Do not setup npcs if there is another battlefield using the same entry npc
+            if self.entryNpc == content.entryNpc then
+                setupEntryNpc = false
+            end
+            if self.exitNpc == content.exitNpc then
+                setupExitNpc = false
+            end
+        end
+    end
+
+    local zoneSection = self.sections[1][self.zoneId]
+    if setupEvents then
+        utils.append(zoneSection, {
+            onEventUpdate =
+            {
+                [32000] = Battlefield.redirectEventUpdate,
+                [32003] = utils.bind(Battlefield.redirectEventCall, "onExitEventUpdate"),
+            },
+            onEventFinish =
+            {
+                [32000] = utils.bind(Battlefield.redirectEventCall, "onEventFinishEnter"),
```

Comment:
We might need to verify and/or move some of the hooks in quests and missions to use this code.  Also, need to verify that other BC's in the same area don't process (I don't think they will looking at code so far, but a definite test case)

Reply:
None of these events are blocking and it will call into other hooks including interactions, zones, and npc files. Consequently other BCs in the same area that are not part of this new system will be using the npc file (such as Burning_Circle.lua) and it'll trigger things twice. This is noted in the "known issues". Easiest solution is to just convert over an entire zone of battlefields at a time. If this is not acceptable then we can come up with another solution.